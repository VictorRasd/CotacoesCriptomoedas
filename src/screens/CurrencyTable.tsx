import React, {Component} from 'react'; 
import { ICurrency } from '../server/TickerServer'
import store from '../index'
import { Box, TableBody, TableCell, TableContainer, TableHead, TableRow, Table, Typography} from '@material-ui/core';
import Select from 'react-select'
import  currencyPairs  from '../server/TickerServer'


interface OwnProps {
}

interface OwnState {
  // Currencies são os dados de cada criptomoeda que serão exibidos na tela
  currencies: ICurrency[]
  // currencyPairSelect é o array que será exibido no select para que o usuário selecione as moedas
  currencyPairSelect: any[]
  // selectedCurrencies são as moedas selecionadas pelo usuário no select
  selectedCurrencies: any[]
}



class CurrencyTable extends Component<OwnProps, OwnState> {
  constructor(props: any) {
    super(props);
    this.state = {
      currencies: new Array<ICurrency>(),
      currencyPairSelect: new Array<any>(),
      selectedCurrencies: new Array<any>()
    }
  }


  componentDidMount(){
    // O ComponentDidMount é utilizado para criar o subscribe na store, buscando assim sempre os dados novos
    this.subscriptionHandler();
  }

  subscriptionHandler(){
    store.subscribe(() => {
      const state = store.getState();
      // O status é 'updated' sempre quando o websocket envia uma nova atualização. O state.data recebido já o array de dados
      // todo atualizado
      if(state.status === 'updated'){
        let currenciesToShow: ICurrency[] = new Array<ICurrency>();
        for( let i = 0 ; i < this.state.selectedCurrencies.length ; i++ ){
          for ( let j = 0 ; j < state.data.length ; j++){
            if(this.state.selectedCurrencies[i].value === state.data[j].name)
              {
                currenciesToShow.push(state.data[j]);
                break;
            }
          }
        }
        this.setState({
          currencies: currenciesToShow,
        })
      }
      // O status 'received' é enviado na primeira busca a API. Dentro desse if é criado todos os campos do select
      // para que somente sejam exibidos criptomoedas que estão sendo trabalhadas na API no momento
      else if(state.status === 'received'){
        let currencyPairsSelect: any = [];
        currencyPairs.forEach(element => {
            let obj: {value: string, label: string};
            obj ={
                value: element,
                label: element
            }
            for ( let i = 0 ; i < state.data.length ; i++){
              if(state.data[i].name === obj.value){
                currencyPairsSelect.push(obj);
                break;
              }
            }
        });
        this.setState({
          currencyPairSelect: currencyPairsSelect
        })
      }
    })
  }

  
  selectCurrencies(event: any){
    if(event === null){
      this.setState({
        selectedCurrencies: new Array<any>()
      })
    }
    else{
      this.setState({
        selectedCurrencies: event
      })
    }
  }

  render() {
    return (
      <Box>
        <Box m={2}>
          <Typography> Favor selecione abaixo as cotações de criptomoedas a serem exibidas </Typography>
        </Box>
        <Box m={2}>
          <Select
            isMulti
            onChange={(event) => this.selectCurrencies(event)}
            name="colors"
            options={this.state.currencyPairSelect}
            className="basic-multi-select"
            classNamePrefix="select"
          />
        </Box>
        <TableContainer>
       <Table size="small" aria-label="sticky table" id="tabela">
         <TableHead>
           <TableRow>
             <TableCell>Nome</TableCell>
             <TableCell align="right">Último</TableCell>
             <TableCell align="right">Menor Pedido</TableCell>
             <TableCell align="right">Maior Oferta</TableCell>
             <TableCell align="right">Mudança Percentual</TableCell>
             <TableCell align="right">Volume Base</TableCell>
             <TableCell align="right">Volume Quotação</TableCell>
             <TableCell align="right">Congelado <br/> ( 0 = Não, 1 = Sim)</TableCell>
             <TableCell align="right">Maior 24hr</TableCell>
             <TableCell align="right">Menor 24hr</TableCell>
           </TableRow>
         </TableHead>
         <TableBody >
           {this.state.currencies.map((currency) => (
             <TableRow key={currency.name} >
               <TableCell component="th" scope="row">
                 {currency.name}
               </TableCell>
               <TableCell align="right">{currency.last}</TableCell>
               <TableCell align="right">{currency.lowestAsk}</TableCell>
               <TableCell align="right">{currency.highestBid}</TableCell>
               <TableCell align="right">{currency.percentChange}</TableCell>
               <TableCell align="right">{currency.baseVolume}</TableCell>
               <TableCell align="right">{currency.quoteVolume}</TableCell>
               <TableCell align="right">{currency.isFrozen}</TableCell>
               <TableCell align="right">{currency.high24hr}</TableCell>
               <TableCell align="right">{currency.low24hr}</TableCell>
             </TableRow>
           ))}
         </TableBody> 
       </Table>
     </TableContainer>
    </Box>
    );
  }
}

export default CurrencyTable;

