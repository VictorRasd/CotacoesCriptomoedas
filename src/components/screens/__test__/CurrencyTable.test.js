import React from 'react'
import ReactDOM from 'react-dom'
import CurrencyTable from '../CurrencyTable.tsx'
import renderer from 'react-test-renderer'
import "@testing-library/jest-dom/extend-expect"

it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<CurrencyTable></CurrencyTable>, div);
})

it("matches snapshot", () => {
    const tree = renderer.create(<CurrencyTable></CurrencyTable>).toJSON();
    expect(tree).toMatchSnapshot()
})

