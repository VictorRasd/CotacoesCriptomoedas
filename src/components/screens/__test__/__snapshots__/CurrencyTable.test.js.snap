// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiBox-root MuiBox-root-4"
>
  <div
    className="MuiBox-root MuiBox-root-5"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
       Favor selecione abaixo as cotações de criptomoedas a serem exibidas 
    </p>
  </div>
  <div
    className="MuiBox-root MuiBox-root-6"
  >
    <div
      className="basic-multi-select css-2b097c-container"
      onKeyDown={[Function]}
    >
      <div
        className="select__control css-yk16xz-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className="select__value-container select__value-container--is-multi css-g1d714-ValueContainer"
        >
          <div
            className="select__placeholder css-1wa3eu0-placeholder"
          >
            Select...
          </div>
          <div
            className="css-b8ldur-Input"
          >
            <div
              className="select__input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-autocomplete="list"
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                disabled={false}
                id="react-select-3-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                spellCheck="false"
                style={
                  Object {
                    "background": 0,
                    "border": 0,
                    "boxSizing": "content-box",
                    "color": "inherit",
                    "fontSize": "inherit",
                    "label": "input",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "1px",
                  }
                }
                tabIndex="0"
                type="text"
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </div>
        </div>
        <div
          className="select__indicators css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className="select__indicator-separator css-1okebmr-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-6q0nyr-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div>
        <input
          name="colors"
          type="hidden"
        />
      </div>
    </div>
  </div>
  <div
    className="MuiTableContainer-root"
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root"
      id="tabela"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
            scope="col"
          >
            Nome
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Último
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Menor Pedido
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Maior Oferta
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Mudança Percentual
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Volume Base
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Volume Quotação
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Congelado 
            <br />
             ( 0 = Não, 1 = Sim)
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Maior 24hr
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
            scope="col"
          >
            Menor 24hr
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
</div>
`;
